### PR TITLE
[EJBCLIENT-309] Add expiration for failed connections in discovery provider

### PR DESCRIPTION
**Pull request by ivagulin**
RFC pull-request to address following scenario:

Steps
1. having cluster of 2 nodes: nodeA and nodeB
2. having ejb-remote client: clientNode
3. nodeB is disabled
4. start clientNode and make some ejb-remote calls (failedConnections will be initialized with nodeB)
5. start nodeB
6. shutdown nodeA
7. make some more ejb-remote calls from clientNode

Actual Result
all calls from ejb-remote node fails with "connection refused". RemotingEJBDiscoveryProvider not try to establish connection to nodeB

Expected Result
ejb calls working. RemotingEJBDiscoveryProvider try to establish connection to nodeB

jira reference
https://issues.jboss.org/browse/EJBCLIENT-309

### CODE REVIEW

**Comment by fl4via** [comment]
@dmlloyd I'm unsure about hard coding timeout instead of using a property to configure it. Wdyt about this, David?

**Comment by dmlloyd** [comment]
Yeah we don't want to hard code values.  Also currentTimeMillis should not be used (use nanoTime instead).

File: src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java [hunk 1]
```diff
@@ -114,6 +113,16 @@ public void removeCluster(final String clusterName) {
...
         effectiveAuthURIs.remove(clusterName);
     }
+    
+    private boolean haveNotExpiredFailedDestination(URI uri) {
+    	if(!failedDestinations.containsKey(uri))
+    		return false;
+    	else {
+    		long failureTimestamp = failedDestinations.get(uri);
+    		long delta = System.currentTimeMillis() - failureTimestamp;
+    		return delta < 5000;
```
Comment (fl4via) [inline src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java]: Please, don't hard code the value of timeout

File: src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java [hunk 2]
```diff
@@ -114,6 +113,16 @@ public void removeCluster(final String clusterName) {
         if (removed != null) removed.clear();
         effectiveAuthURIs.remove(clusterName);
     }
+    
+    private boolean haveNotExpiredFailedDestination(URI uri) {
+    	if(!failedDestinations.containsKey(uri))
+    		return false;
+    	else {
+    		long failureTimestamp = failedDestinations.get(uri);
+    		long delta = System.currentTimeMillis() - failureTimestamp;
```
Comment (dmlloyd) [inline src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java]: This should be using `nanoTime`.

**Comment by ivagulin** [comment]
- issue added to pull-request summary
- review comments fixed

**Pull request merged**